proposals/unsigned/unsigned.txt

```diff
@@ -0,0 +1,228 @@
+
+To perform mixed-mode arithmetic with INTEGER or REAL values,
+the UNSIGNED operand must be converted to an INTEGER or REAL
+value explicitly via the INT or REAL intrinsic functions.
+
+
+# 3. Avoiding traps and pitfalls
+
+There are numerous well-known traps and pitfalls when using unsigned
+integers. We attempt to avoid these as follows:
+- comparison of signed vs. unsigned values: require conversion via
+  an intrinsic function or other means.
+- overflow from assignment of large UNSIGNED values to similar-sized
+  INTEGER entities: Either accept truncation or specify the KIND with a
+  larger range to the INT intrinsic function.
+- confusion about modulo arithmetic, especially with respect to
+  subtraction (e.g., 3u - 5u < 3u .EQV. .false.) is avoided
+  because `3u - 5u` is undefined and compilers can optionally give a
+  compile-time or runtime error.
+
+
+# 4. Proposal
+
+- A type name tentatively called UNSIGNED, with the same KIND
+  mechanism as for INTEGER, plus a SELECTED_UNSIGNED_KIND function,
+  is added to implement unsigned integers.
+
+- Unsigned integer literal constants are marked with a U suffix,
+  with an optional KIND specifier attached via the usual underscore.
+
+- Add a conversion function UINT, with an optional KIND.
+
+- Prohibit binary operations between INTEGER and UNSIGNED or
+  REAL and UNSIGNED without explicit conversion.
+
+- Permit unsigned integer values in a SELECT CASE.
+
+- Prohibit unsigned integers as index variables in a DO statement
+  or as array indices.
+
+- Allow unsigned integers to be read or written in list-directed,
+  namelist or unformatted I/O, and by using the usual edit
+  descriptors such as I, B, O and Z.
+
+- Allow UNSIGNED arguments to some intrinsics:
+    - BGE(UNSIGNED, UNSIGNED) and friends
+    - BIT_SIZE(UNSIGNED)
+    - BTEST(UNSIGNED, INTEGER)
+    - DIGITS(UNSIGNED)
+    - DSHIFTL(UNSIGNED, UNSIGNED, INTEGER)
+    - DSHIFTR(UNSIGNED, UNSIGNED, INTEGER)
+    - HUGE(UNSIGNED)
+    - IAND(UNSIGNED, UNSIGNED), IEOR, IOR, NOT
+    - IBCLR(UNSIGNED, INTEGER), IBITS, IBSET
+    - ISHFT(UNSIGNED, INTEGER, INTEGER) and ISHFTC
+    - LEADZ(UNSIGNED) and TRAILZ
+    - MERGE_BITS(UNSIGNED, UNSIGNED, UNSIGNED
+    - MIN(UNSIGNED, ...) and MAX
+    - MOD(UNSIGNED, UNSIGNED) and MODULO
+    - MVBITS(UNSIGNED, INTEGER, INTEGER, UNSIGNED, INTEGER)
+    - POPCNT(UNSIGNED) and POPPAR
+    - RANGE(UNSIGNED)
+    - SHIFTA(UNSIGNED, INTEGER), SHIFTL, SHIFTR
+    - TRANSFER(UNSIGNED, UNSIGNED, INTEGER)
+
+- Allow UNSIGNED arguments to some array intrinsics:
+    - IALL(UNSIGNED array, INTEGER, [, mask]) and friends
+    - IPARITY(UNSIGNED array, INTEGER [, mask])
+    - CSHIFT(UNSIGNED array, INTEGER, INTEGER)
+    - DOT_PRODUCT(UNSIGNED array, UNSIGNED array)
+    - EOSHIFT(UNSIGNED array, INTEGER, INTEGER)
+    - FINDLOC(UNSIGNED array, UNSIGNED, ...)
+    - MATMUL(UNSIGNED array, UNSIGNED array)
+    - MAXLOC(UNSIGNED array, ...), and MINLOC
+    - MAXVAL(UNSIGNED array, ...), MINVAL
```

Comment:
SUM, PRODUCT

PACK, RESHAPE, SPREAD, TRANSPOSE, UNPACK

SELECTED_UNSIGNED_KIND

UINT

Reply:
Thanks! Added in af6dd673fa77cfcae4bb107375a823aff2ca6629, arguments must still be added.